### ISSUE TITLE
Test data being used for validation data?

### ISSUE THREAD

**Issue by KeironO** [fchollet/keras#1753]
Hey there,

In many of your examples you seem to be using the test data as validation data, wouldn't this create an overfitted model that is unable to generalise?

Example: https://github.com/fchollet/keras/blob/master/examples/cifar10_cnn.py#L80

**Comment by gw0** [comment]
Validation data is not used for training (or development of the model). Its purpose is to track progress through validation loss and accuracy.

**Comment by tboquet** [comment]
This is just a terminology used, it could also be called `X_val` and `y_val`. You would have to use another set of data (often called the test set) if you want to do some hyperparameters optimization.

**Comment by KeironO** [comment]
No, no, no, no. This is incorrect.

The validation data needs to be separate to both the training and test data. 

The results that this model gives are false as they are built to over-fit on the data, something that we shouldn't be suggesting is right.

Instead we should opt to use ```validation_split``` method in ```model.fit```, before we confuse people further.

**Comment by tboquet** [comment]
I think you are confused about the naming of the datasets. The results of each example are correct and the models are not overfitted since the validation data is not used in the fitting process. You just have less control using `validation_split`.

Take a look at those lines:
[data fed to the `_fit` method using `validation_data`](https://github.com/fchollet/keras/blob/master/keras/models.py#L615)
[data fed to the `_fit` method using `validation_split`](https://github.com/fchollet/keras/blob/master/keras/models.py#L615)
[evaluation of the validation data](https://github.com/fchollet/keras/blob/master/keras/models.py#L294-L296)

You could see that it's the same treatment for the data you pass in the `_fit` method if you choose `validation_split` or `validation_data`. Then you also see that this data is not passed to the `_train` method so the optimizer doesn't update the parameters with respect to this data.

**Comment by KeironO** [comment]
Not quite what I meant, take a look here - this is directly taken from cifar10_cnn in the examples folder.
```
if not data_augmentation:
    print('Not using data augmentation.')
    model.fit(X_train, Y_train, batch_size=batch_size,
              nb_epoch=nb_epoch, show_accuracy=True,
              validation_data=(X_test, Y_test), shuffle=True)
else:
    print('Using real-time data augmentation.')

```

Surely this is an incorrect method of training our models, as the entire point is to produce models that are able to generalise in classification tasks - and by validating on the test dataset then the network will have already seen it, making the evaluation of the models incorrect.

**Comment by tboquet** [comment]
Ok I don't think I understand your point, so let's summarize:

Are you saying that the the validation data should be fed to the `fit` method like this?
```python
if not data_augmentation:
    print('Not using data augmentation.')
    model.fit(X_train, Y_train, batch_size=batch_size,
              nb_epoch=nb_epoch, show_accuracy=True,
              validation_data=(X_val, Y_val), shuffle=True)
else:
    print('Using real-time data augmentation.')
```

**Comment by KeironO** [comment]
Yes.

The entire point of cross validation is that the network trains on training data, validates the training using a separate validation test set and then evaluates the performance of the model on another separate test set.

These networks are unable to effectively generalise, as they have already seen the test data.

**Comment by KeironO** [comment]
Another comment was made about this here https://github.com/fchollet/keras/issues/916 where Francios stated it was just an example of

> real-time monitoring

What on earth does that even mean?

**Comment by neggert** [comment]
I think this is just a nomenclature issue. The correct way to train a net is to have 3 data sets: train, validation, and test. The training set is obvious. The validation set is checked during training to monitor progress, and possibly for early stopping, but is never used for gradient descent. The test dataset is the best measure of the network accuracy, and should only be used once, once all training is finished.

So I think the issue is that, strictly speaking, the dataset in that cifar example is a validation dataset, not a test dataset.

**Comment by KeironO** [comment]
My point is that it's extremely misleading. Take ```mnist_cnn.py``` for example.

```python
model.fit(X_train, Y_train, batch_size=batch_size, nb_epoch=nb_epoch,
          show_accuracy=True, verbose=1, validation_data=(X_test, Y_test))
score = model.evaluate(X_test, Y_test, show_accuracy=True, verbose=0)
```

The header of the python file claims that the model is able to achieve:

> 99.25% test accuracy after 12 epochs

Well of course its able to, it's badly over-fitted!

**Comment by fchollet** [comment]
You seem really confused. I don't think this discussion is going anywhere. I suggest you to go back to the basics of machine learning, and not to put so much weight on slight discrepancies between naming conventions in your textbook and in the real world.

**Comment by KeironO** [comment]
I think you're missing the point entirely. I couldn't care less to what you decide to call your variables - call then kitty, dog and spider all I care.

The problem is that I believe that a few of the examples are misleading. Look at ```mnist_cnn.py``` for example, you perform validation on the test set and then evaluate its performance after the run.

This model has been trained and tested incorrectly, and for people attempting to get to grips with things is a little bit confusing.

**Comment by jgc128** [comment]
I agree with KeironO. The point is that in these examples test data is used as validation data and this is wrong.

For example:

>99.25% test accuracy after 12 epochs

How did the number 12 appear? After 12 epochs the model showed a worse performance on the validation data? In this case it literally means "train the model until the performance on test data starts to go down".

**Comment by gw0** [comment]
> The entire point of cross validation ...

@KeironO Those examples do **not** use cross validation, just a classic train/test data split. Consequently no actual overfitting happened, in the worst case it was used in the sense @neggert or @jgc128 described -- to monitor progress to set the `nb_epochs = 12`. But maybe this did not happen, the author of the example did not monitor progress to manually tweak `nb_epochs` and just left it at value 12 from the beginning. Strictly speaking in the first case it is a validation dataset used for manual tweaking and in the second it is a test dataset.

**Comment by bayesrule** [comment]
@KeironO  take a close look at what @neggert said : "**The validation set** is checked during training to monitor progress, and possibly for early stopping, **but is never used for gradient descent.**"

Indeed, the data set fed to the argument "validation_data" of model.fit() in Keras is never used for training. Using test data for this validation is SAFE.

**Comment by bbartoldson** [comment]
@fchollet, I was wondering why there was no third dataset too. I expected training, validation, and test datasets. As it's currently written, [mnist_cnn.py](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py#L69) puts the test data into the validation data slot:

```
model.fit(X_train, Y_train, batch_size=batch_size, nb_epoch=nb_epoch,
          verbose=1, validation_data=(X_test, Y_test))
```
which will confuse some people like @KeironO and risk leading others astray. For example, someone could make a change to [mnist_cnn.py](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py#L69), run the program, and decide whether or not to keep their model changes based on the change in the test dataset accuracy. A novice may even be nudged toward that pitfall since the tutorial program explicitly states that "there is still a lot of margin for parameter tuning". As the TensorFlow MNIST tutorial and Yann LeCun suggest (see below), a separate validation dataset should be used for tuning purposes, not the test data.

[Tensorflow's MNIST tutorial](https://www.tensorflow.org/versions/master/tutorials/mnist/beginners/index.html) uses 5,000/60,000 training images as a validation set:

> The downloaded data is split into three parts, 55,000 data points of training data (mnist.train), 10,000 points of test data (mnist.test), and 5,000 points of validation data (mnist.validation). This split is very important: it's essential in machine learning that we have separate data which we don't learn from so that we can make sure that what we've learned actually generalizes!

Yann LeCun, [who helped to develop the MNSIT dataset](http://yann.lecun.com/exdb/mnist/), and others came up with a state-of-the-art result on MNIST in the paper [What is the Best Multi-Stage Architecture for Object Recognition?](http://yann.lecun.com/exdb/publis/pdf/jarrett-iccv-09.pdf), in which they discuss their use of a validation set:

> experiments were run on the MNIST dataset, which contains 60,000 gray-scale 28x28 pixel digit images for training and 10,000 images for testing... A validation set of size 10,000 was set apart from the training set to tune the only hyper-parameter: the sparsity constant λ... The system was trained with a form of stochastic gradient descent on the 50,000 non-validation training samples until the best error rate on the validation set was reached (this took 30 epochs). It was then tuned for another 3 epochs on the whole training set.

Granted, [mnist_cnn.py](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py#L69) does not violate any machine learning principles I'm aware of because it's not tuning a parameter or the model by checking the test data. Nevertheless, the tutorial is currently structured so that someone might adopt that incorrect practice, and the tutorial has clearly confused some people already. 

All of this said, I propose modifying the tutorial program to use `validation_split`, which is described in the [Keras documentation](http://keras.io/models/sequential/) as:

> validation_split: float (0. < x < 1). Fraction of the data to use as held-out validation data.

I suggest using the same amount of validation data as the TensorFlow tutorial, which is 5,000/60,000 images, or 1/12. I tested the program using this modification and obtained 99.00% test data accuracy. I then ran 3 more epochs using all 60,000 training images, which is the process that LeCun used and described above, and reached 99.09% accuracy. This can be seen in the picture below. 

![mnist_cnn](https://cloud.githubusercontent.com/assets/15717529/15802542/a931cba0-2a83-11e6-9ddb-6a13d9d471e1.png)

A pull request to incorporate this change is #2899. Please let me know if you have any questions, or if you would like me to make a modification.

**Comment by talpay** [comment]
The example is technically correct but only until someone starts to optimize hyperparameters manually without reimplementing the `fit` call to use `validation_split`. In that case people *will* overfit to the test data. So why not directly implement the split (like @bbartoldson suggests) as this is guaranteed to confuse absolutely noone and reduce the risk of students/novices adapting bad practices.

**Comment by ygorcanalli** [comment]
I agree with @talpay, i'ts important to make the example didactic for novices, who probably will want to optimize hyperparameters (like the code comments suggests). Just adding the parameter `validation_split` on `fit` method will solve the problem, like @fchollet shows.

**Comment by lbugnon** [comment]
I think the model.fit method (and documentation) is misleading. It's not that obvious that the "validation_data=(X_test, Y_test)" is used only for documenting progress and none optimization is made using that data. In fact, if you are going to feed a "training" function like fit (i know, its not the training function, but similar) with a train and test set, you should state CLEARLY why are you doing that, and what use is given to this test(or validation) data.

**Comment by SarthakYadav** [comment]
I've never raised this issue, but actually, this is something I have just been ignoring in all the Keras examples. There is a reason elaborate cross-validation techniques were developed. If you could just train a model and evaluate it's performance on the test set, all the heaps and bounds of research done should be deemed useless: we have access to both training and test data. Maybe some one just used up the test set itself!!!. Actual test of the model is GENERALIZATION, not accuracy numbers on the train or even validation set to some extent. If the same methodology was adopted in all research work (since the issue arised with CIFAR and MNIST sets in mind, let's take their example), i.e we were continuously updating Network parameters based on the test set acting as a validation set, we aren't training a model that generalizes well, we are creating a model that generalizes well ON THE TEST SET ITSELF. Let's assume that a model you trained on CIFAR10 in this way then encounters new unseen data, which should be the purpose of the model in the first place: I don't believe the performance will be the same.

The reason I have been ignoring this, is because I know that the sole purpose of an example is to demonstrate how the API is capable of training time validation, and I know the difference AND THE NEED for separate validation set. Also, in all the examples ever presented in keras documentation, NO OVERFITTING EVER HAPPENED, since we didn't update the hyperparameters, and the validation data is not used by the network being trained.
 
So the issue itself is not at all that in vain. It can be misleading. (as it has proven to be for KeironO).

**Comment by harikrishnavydana** [comment]
@fchollet  why there is a test folder in keras code..what is the use

**Comment by gw0** [comment]
@harikrishnavydana You mean [unit tests](https://github.com/fchollet/keras/tree/master/tests) for Keras?

Otherwise, stop posting non-relevant comments to closed issues!!!

**Comment by harikrishnavydana** [comment]
yes i was asking for unit tests @gw0

**Comment by Euphemiasama** [comment]
Is monitoring the validation loss to reduce the learning rate when it reaches a plateau not considered tweaking the parameters ? So it that case aren't we peeking into the test data (validation data is the case of keras) which would eventually lead to overfitting?

**Comment by pGit1** [comment]
@bbartoldson and @KeironO  are 100% correct, period. However these small discrepancies should be ignored in the examples as they are just examples to demonstrate how fitting with a validation set works. In real life having access to a test set and using it to inform model architectures and other hyper parameters decisions leads to overfitting guaranteed. This is common ML knowledge and detracting opinions are simply/utterly wrong...the evidence at this point in ML is daunting.

That said I dont think the focus of the examples is to show the pitfalls of overfitting due to improper validations schemes. That should be well understood prior to building any models, nevermind powerful function approximators that can memorize noise in data.

**Comment by mohapatras** [comment]
The reason @fchollet used test data in validation is correct in a way. If we look into the basics of machine learning "Validation is a process of tracking record of loss and accuracy" which will help us tuning and choosing better hyper parameters. So in a way in the mnist_cnn.py example for validation calculation test set passed and used. 

The only confusion that occurs is to the beginners. As it has been taught and is usually correct that test data should not be used in case of measuring anything as the real problem arises when you adjust hyperparameters such as learning rate based on the validation scores. By doing this you have incorporated validation data into your model and the scores are no longer independent.
Therefore, If you want an unbiased estimate of the final score you would need a separate test set.
Also @bbartoldson @gw0 's opinion about the subject matter is correct.

**Comment by grafael** [comment]
So, after this long discussion... If I have 3 datasets: (train, valid and test) how can I explicitly tell to Keras to use those?  Is it right the below snippet? I've been using this kind of approach for several months.  @fchollet @mohapatras 

```python
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=15,
          validation_data=(x_valid, y_valid))

score, acc = model.evaluate(x_test, y_test,
                            batch_size=batch_size)
```

**Comment by pGit1** [comment]
@Rafael,

Nothing wrong with that at all. As long as you don't use score,acc to
update your hyper-parameters. I mean you can but it may give you an overly
optimistic view.

**Comment by TohnoYukine** [comment]
Improper use of test set is improper. I think using that validation_split parameter is also easier for beginners to learn than using test data set as validation set. Beginners will really be confused at that misuse of test set as validation set, especially when they dive into hyper parameter. Please update the code before it misleads more people.

**Comment by ygorcanalli** [comment]
My vote is with @TohnoYukine, would be a best approach for beginners to use validation_split.

**Comment by joeantol** [comment]
Perhaps the Keras documentation will provide some clarity:

**validation_split**: Float between 0 and 1. Fraction of the training data to be used as validation data. The model will set apart this fraction of the training data, will not train on it, and will evaluate the loss and any model metrics on this data at the end of each epoch. The validation data is selected from the last samples in the x and y data provided, before shuffling.

**validation_data**: tuple (x_val, y_val) or tuple (x_val, y_val, val_sample_weights) _on which to evaluate the loss and any model metrics at the end of each epoch_. The model **will not be trained** on this data. This will override  validation_split.

(emphasis mine)

**Comment by rexplorations** [comment]
Instructive discussion as I am resolving the same kind of issue with a project I'm on. Using [testX, testY] in the validation set does not affect the gradient descent (it is only done on the training set). And I get the same results for gradient descent if I don't provide a validation_data argument, assuming I choose to run the model past the test sets separately.

**Comment by rfernand2** [comment]
Good discussion on a often misunderstood topic.  These example programs like MNIST are important because many developers new to ML use them as the basis of their actual applications, so we should strive to reflect best practices as space/time permits.  

@AnanyaKumar makes a good point about the splits being used at different time scales.  The best practice here is to use your train and validation (aka dev) sets for running all of your experiments, trying different hyperparameters, architectures, etc.  Then, when you are done experimenting and have chosen a model and all its hyperparmeters, and you are ready to "spend" your holdout test set on a single, one-time evaluation, you combine your train and validation set into a single training set and do a final training and then, you evaluation of your test set and report it's accuracy.  

To support the above, I propose that the majority of Keras examples like MNIST be focused on running training experiments with train and validation data splits, reporting their final results as "validation accuracy" (or whatever metric is being reported).  In addition, maybe there can be 1 or more examples that show how to do the final training and test evaluation, clearly labeled as such.

**Comment by baldesco** [comment]
> The best practice here is to use your train and validation (aka dev) sets for running all of your experiments, trying different hyperparameters, architectures, etc. Then, when you are done experimenting and have chosen a model and all its hyperparmeters, and you are ready to "spend" your holdout test set on a single, one-time evaluation, you combine your train and validation set into a single training set and do a final training and then, you do an evaluation of your test set and report it's accuracy as such.

I think this sums up pretty well the confusion generated by the keras examples. For model development and hyperparameter tuning, train on `train set` and validate on `validation set`. For the final model, train on `train+validation set` and validate on `test set`.

**Comment by StoneCypher** [comment]
Any training based on a validation set is deeply corrupt and fails both rudimentary and practical 101 principles.

The number of people calling this a nomenclature issue is bewildering.  This is basic practice.  Under no circumstance is including a validation set a "best practice."

That defies the core concept of a validation set.  As soon as it has been included in the training in any way, it is no longer a validation set, and you have no correct mechanism for validation.

This is easily tested.  Just do the same job twice, by making two validation sets.  One is your standard validation set.  The other is a set for validating your development approach.

Now, set the second validation set aside.  Do the same training in two batches of ten, once with the first validation set included, the second time without.

Let those run with whatever automatic hyperparameter tuning you want, ***as long as it's the same kind for both batches***.  No manual hyperparameter tuning, because then you might accidentally be doing a better job on one than the other, without intent.

Next, when both batches have given the same amount of work to every item inside, call a stop.

Now, use the second validation set to see which items in the batches did a better job: the ones with the first validation set included, or the ones without?

You're going to see dramatically, ***embarrassingly*** more overfitting in the batch including validation than the batch not.

I'm really pretty horrified that, years later, this is still closed incorrectly.  Someone should have stepped in by now.  Students have been making decisions on the basis of the names in this thread for years, and if you go to any college professor worth their salt and ask this question carefully to not imply either answer, 95% of them are going to disagree with the way this thread plays out.

It is *not* the case that this thread has it right.

**Comment by talpay** [comment]
> Edit: there are apparently many more people who are confused about the purpose and usage of validation data and test data in machine learning. If you need a textbook to learn about these concepts, please see my book [Deep Learning with Python](https://www.manning.com/books/deep-learning-with-python).

Every single person in this thread understands how the split process works. Around half of the people in this thread understand the intention behind the code and are saying that it is **misleading** and can **lead** to incorrect usage very easily (evidenced by the other half being confused about the code). Even after 4 years, you still seem to be reading this as people saying the code is technically **incorrect** and people therefore not understanding the basics. With many posts going so much into depth as to explain their position and you still misconstruing their positions and giving condescending arguments like "git gud, read my book", it really feels like you're just acting in bad faith and are more concerned with "being right" than maximizing the amount of people that benefit from the code.
